RocketChat/Rocket.Chat
# [FIX] Client disconnection on network loss

## Proposed changes (including videos or screenshots)

Agent gets disconnected (or Unregistered) from asterisk in multiple ways. The goal is that agent should remain online
unless agent explicitly logs off.
Agent can stop receiving calls in multiple ways due to network loss. Network loss can happen in following ways.
1. User tries to switch the network. User experiences a glitch of disconnectivity. This can be simulated by turning the network off
in the network tab of chrome's dev tool. This can disconnect the UA if the disconnection happens just before the registration refresh.
2. Second reason is when computer goes in sleep mode.
3. Third reason is that when asterisk is crashed/in maintenance mode/explicitly stopped.

Solution:
The idea is to detect the network disconnection and start the start the attempts to reconnect.
The detection of the disconnection does not happen in case#1. The SIPUA's UserAgent transport does not
call onDisconnected when network loss of such kind happens. To tackle this problem, window's online and offline event handlers are
used.

The number of retries is configurable but ideally it is to be kept at -1. Whenever disconnection happens, it should keep on trying to
reconnect with increasing backoff time. This behaviour is useful when the asterisk is stopped.

When the server is disconnected, it should be indicated on the phone button.


## Issue(s)
https://app.clickup.com/t/245c0d8

It is ambiguous on what should happen when the disconnection happens while the agent is on call.
## Steps to test or reproduce
There are atleast 4 known ways to reproduce the problem
1. Change the network. e.g switch from home broadband to tethered network.
2. Go to chrome's DevTools -> Network and change No Throttling to offline.
3. Put the computer to sleep.
4. Stop asterisk

## Further comments
Default expiration time of Asterisk is 600 seconds. So unregistration/re-registration will happen only after 10 minutes.

There are several test-cases. This feature should be tested in following scenarios.
Testcase#1
1. Turn off the local network. 
2. Make a call to the extension.
3. Verify that the network is actually disconnected by verifying that the UA does not receive the call.
4. Bring back the network.
5. Verify on asterisk that the userAgent is connected by running command pjsip show endpoint xxxx. In the result of this command verify that the extension status is "not in use" and NOT "unavailable"

Testcase#2.
1. Verify steps 2 onwards by  changing DevTools -> Network  No throttle to offline.
2. Verify step 2 onwards by switching the network.

Testcase#3
1. Put computer on sleep.
2. Make a call to the extension from another device.
3. Verify that the agent is actually disconnected by verifying that the UA does not receive the call.
4. Wake up the computer.
5. Verify on asterisk that the userAgent is connected by running command pjsip show endpoint xxxx. In the result of this command verify that the extension status is "not in use" and NOT "unavailable"

Tasecase#4.
1. Stop Asterisk.
2. Make a call to the extension from another device.
3. Verify that the asterisk  is actually stopped by verifying that the UA does not receive the call.
4. Restart asterisk.
5. Verify on asterisk that the userAgent is connected by running command pjsip show endpoint xxxx. In the result of this command verify that the extension status is "not in use" and NOT "unavailable"

Testcase#5
1. Goto Administrator -> Call Center -> Server Configuration.
2. Change retry count to 2.
3. Stop the asterisk.
4. Let the code retry for 2 times.
5. Make sure that after 2 retries, a proper error is shown as a tooltip of the phone button.
6. Make sure that UA does not receive any calls.
